### PR TITLE
CLI and config file option cleanup and harmonization

### PR DESCRIPTION
**Pull request by dcbw**
1. harmonizes CLI arguments between the overlays, ovnkube, and gateway
2. moves general config handling (logging, Kube API auth, and OVN NB/SB auth) to config.go
3. adds all general config options to the config file, renames existing config options to match CLI arguments, and moves them to discrete sections
4. composes the final configuration from the CLI options, config file, and finally defaults, in that order.
5. updates the Vagrant config to use new options

example config file:
```
[default]
mtu=1500
conntrack-zone=64321

[kubernetes]
kubeconfig=/path/to/kubeconfig
apiserver=https://1.2.3.4:6443
token=<blahblahblah>
cacert=/path/to/some/cert.crt

[logging]
loglevel=5
logfile=/var/log/ovnkube.log

[cni]
conf-dir=/etc/cni/net.d
plugin=ovn-k8s-cni-overlay

[ovnnorth]
address=ssl://1.2.3.4:6641
client-privkey=/path/to/private.key
client-cert=/path/to/client.crt
client-cacert=/path/to/client-ca.crt
server-privkey=/path/to/private.key
server-cert=/path/to/server.crt
server-cacert=path/to/server-ca.crt

[ovnsouth]
<same as ovn north>
```

@shettyg @rajatchopra

### CODE REVIEW

File: go-controller/pkg/config/config.go [hunk 1]
```diff
@@ -1,105 +1,575 @@
...
+		return nil, fmt.Errorf("failed to parse OVN DB URL %q: %v", urlString, err)
+	}
+	host, port, err := net.SplitHostPort(url.Host)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse OVN DB host/port %q: %v", url.Host, err)
+	}
+	// OVN requires the --db argument to be an IP, not a DNS name
+	ip := net.ParseIP(host)
+	if ip == nil {
+		return nil, fmt.Errorf("OVN DB host %q must be an IP address, not a DNS name", url.Host)
```
Comment (pecameron) [inline go-controller/pkg/config/config.go]: Would it be possible to just resolve the name to an IP for the user?

Comment (dcbw) [reply go-controller/pkg/config/config.go]: @pecameron it might, but we'd have to do that internally.  This isn't a regression though, this statement was present before this PR.

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 2]
```diff
@@ -20,88 +20,179 @@ import (
...
+		},
+		cli.StringFlag{
+			Name:  "ovn-north-client-cacert",
+			Usage: "CA certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.",
+		},
+
+		// OVN southbound database options
+		cli.StringFlag{
+			Name:  "ovn-south-db",
+			Usage: "IP address and port of the OVN southbound API (eg, ssl://1.2.3.4:6641).  Leave empty to use a local unix socket.",
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: @pecameron fixed

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 3]
```diff
@@ -20,88 +20,179 @@ import (
...
+			Usage: "A CIDR notation IP range from which k8s assigns " +
+				"service cluster IPs. This should be the same as the one " +
+				"provided for kube-apiserver \"-service-cluster-ip-range\" " +
+				"option.",
+		},
 
+		// OVN northbound database options
+		cli.StringFlag{
+			Name:  "ovn-north-db",
+			Usage: "IP address and port of the OVN northbound API (eg, ssl://1.2.3.4:6641).  Leave empty to use a local unix socket.",
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: Yeah, this is just an example.  TCP is still allowed.

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 4]
```diff
@@ -20,88 +20,179 @@ import (
...
+			Name:  "ca-cert",
+			Usage: "CA cert for the Kubernetes api server",
+		},
+		cli.StringFlag{
+			Name:  "token",
+			Usage: "Bearer token to use for establishing ovn infrastructure",
+		},
+		cli.StringFlag{
+			Name:  "cluster-subnet",
+			Value: "11.11.0.0/16",
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: This was taken from the previous code's default value:

`-	clusterSubnet := flag.String("cluster-subnet", "11.11.0.0/16", "Cluster wide IP subnet to use")`

**Comment by dcbw** [comment]
@shettyg @rajatchopra thoughts on this one?

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 5]
```diff
@@ -221,7 +132,7 @@ func runOvnKube(ctx *cli.Context) error {
 	}
 
 	// creates the clientset
-	clientset, err := kubernetes.NewForConfig(config)
+	clientset, err := kubernetes.NewForConfig(kconfig)
```
Comment (alinbalutoiu) [inline go-controller/cmd/ovnkube/ovnkube.go]: It seems to be the same as before, it should be possible to use NewClientset here too.

File: go-controller/pkg/config/config.go [hunk 6]
```diff
@@ -1,105 +1,589 @@
...
+		}
+		logrus.Infof("Parsed config file %s", f.Name())
+		logrus.Infof("Parsed config: %+v", cfg)
+
+		if defaults != nil {
+			fillDefaults(&cfg, defaults)
+		}
+	}
+
+	// command-line > config file > default values
```
Comment (alinbalutoiu) [inline go-controller/pkg/config/config.go]: I think using global variables for the default values would be easier to read and understand in my opinion and all default values will be in only one place. Or at least a structure with all the default values defined would work as well.

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 7]
```diff
@@ -137,81 +233,105 @@ func main() {
...
 	}
-	if *netController {
+	if netController {
 		ovnController := ovn.NewOvnController(clientset, factory)
-		ovnController.NodePortEnable = *nodePortEnable
+		ovnController.NodePortEnable = nodePortEnable
 		ovnController.Run()
 	}
-	if *master != "" || *netController {
+	if master != "" || netController {
```
Comment (alinbalutoiu) [inline go-controller/cmd/ovnkube/ovnkube.go]: This is unrelated to this PR.

Why do we run forever if we do only a master-init? Shouldn't it run forever only if the netController is specified as well? Like the following:
```
master != "" && netController
```

At the moment ovnkube returns only if called without master name and netController flag. In this case it does almost nothing useful. In my opinion there should be a warning at least.

Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: @alinbalutoiu in the 'master' and netController roles, it needs to run forever.  The master code has the node watch to allocate CIDRs to nodes, and the netController handles policy and such.  They both need to continue running.

At the moment, only the node can be run & quit.

**Comment by rajatchopra** [comment]
This is looking good. Separately, it seems we are closer to deprecating the ovn-k8s-overlay maybe?

File: go-controller/pkg/config/config.go (hunk 6 above)
Comment (rajatchopra) [reply go-controller/pkg/config/config.go]: Probably a complete example file with all the options listed would be useful.

Comment (rajatchopra) [reply go-controller/pkg/config/config.go]: Copy from the PR description itself above and put it somewhere along the README

File: go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go [hunk 8]
```diff
@@ -148,39 +149,20 @@ func cmdAdd(args *skel.CmdArgs) error {
...
-		k8sAPIServer = fmt.Sprintf("http://%s", k8sAPIServer)
-		cfg, err = clientcmd.BuildConfigFromFlags(k8sAPIServer, "")
+		err = fmt.Errorf("invalid kubernetes configuration")
 	}
 	if err != nil {
 		return fmt.Errorf("Failed to get kubeclient: %v", err)
 	}
 
-	clientset, err := kubernetes.NewForConfig(cfg)
+	clientset, err := kubernetes.NewForConfig(kubeCfg)
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go]: @alinbalutoiu I actually do this in a subsequent commit "go-controller: consolidate common kubeclient creation code".  If that's not what you're thinking of, let me know.

File: go-controller/cmd/ovnkube/ovnkube.go (hunk 5 above)
Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: @alinbalutoiu I actually do this in a subsequent commit "go-controller: consolidate common kubeclient creation code". If that's not what you're thinking of, let me know.

Comment (alinbalutoiu) [reply go-controller/cmd/ovnkube/ovnkube.go]: @dcbw yep it's fine, you sent the new commits while I was reviewing the patches. This review got marked automatically as outdated.

File: go-controller/cmd/ovnkube/ovnkube.go (hunk 7 above)
Comment (alinbalutoiu) [reply go-controller/cmd/ovnkube/ovnkube.go]: Thanks a lot for explaining it!

**Comment by alinbalutoiu** [comment]
Looks good to me, the only thing left might be as @rajatchopra suggested is to have the config file example copied somewhere in README.

File: go-controller/pkg/config/config.go (hunk 6 above)
Comment (dcbw) [reply go-controller/pkg/config/config.go]: @alinbalutoiu this only works for a couple of the simple structs; for others (Kubernetes and OvnAuth) we need to construct the defaults based on things like whether the address is HTTP or HTTPS, or we need to grab the default value from OVS external-ids.

I'll do some reworking to make as many defaults explicitly specified as we can.

Comment (alinbalutoiu) [reply go-controller/pkg/config/config.go]: @dcbw I was just thinking of having the default values defined as constants, for example:
```
Default.MTU = getConfigInt(cliConfig.Default.MTU, cfg.Default.MTU, 1400)
```
Instead of having that 1400 over there to be defined as a constant in the beginning of the file.
Anyway that is just a thought, the way you did it works as well.

File: go-controller/etc/ovn_k8s.conf [hunk 9]
```diff
@@ -1,15 +1,18 @@
...
-unix-socket=/var/run/openvswitch/ovnnb_db.sock
-cni-conf-path=/etc/cni/net.d
-cni-link-path=/opt/cni/bin/
+
+[Logging]
+logfile=/var/log/openvswitch/ovn-k8s-cni-overlay.log
+
+[CNI]
+cni-conf-dir=/etc/cni/net.d
 cni-plugin=ovn-k8s-cni-overlay
```
Comment (alinbalutoiu) [inline go-controller/etc/ovn_k8s.conf]: Can we have a place where we can see all the possible config values? Either somewhere in README or here would be great. You gave an example in the PR description, but I don't see any other way of finding all the possible config options besides the PR description or looking through the code.

Comment (dcbw) [reply go-controller/etc/ovn_k8s.conf]: @alinbalutoiu will do

**Comment by shettyg** [comment]
@dcbw Were you planning to add the readme in this series?

**Comment by rajatchopra** [comment]
Merging this PR to unblock other work.

**Comment by dcbw** [comment]
@rajatchopra @alinbalutoiu @shettyg yeah, was planning on updates for readme and testcases, but this got merged in the mean time :)  I'll do another PR for those.
